src/services/codefixes/disableJsDiagnostics.ts

```diff
@@ -1,18 +1,47 @@
 /* @internal */
 namespace ts.codefix {
-    registerCodeFix({
-        errorCodes: getApplicableDiagnosticCodes(),
-        getCodeActions: getDisableJsDiagnosticsCodeActions
+    const groupId = "disableJsDiagnostics";
+    const errorCodes = mapDefined(Object.keys(Diagnostics), key => {
+        const diag = (Diagnostics as MapLike<DiagnosticMessage>)[key];
+        return diag.category === DiagnosticCategory.Error ? diag.code : undefined;
     });
 
-    function getApplicableDiagnosticCodes(): number[] {
-        const allDiagnostcs = <MapLike<DiagnosticMessage>>Diagnostics;
-        return Object.keys(allDiagnostcs)
-            .filter(d => allDiagnostcs[d] && allDiagnostcs[d].category === DiagnosticCategory.Error)
-            .map(d => allDiagnostcs[d].code);
-    }
+    registerCodeFix({
+        errorCodes,
+        getCodeActions(context) {
+            const { sourceFile, program, newLineCharacter, span } = context;
 
-    function getIgnoreCommentLocationForLocation(sourceFile: SourceFile, position: number, newLineCharacter: string) {
+            if (!isInJavaScriptFile(sourceFile) || !isCheckJsEnabledForFile(sourceFile, program.getCompilerOptions())) {
+                return undefined;
+            }
+
+            return [{
+                description: getLocaleSpecificMessage(Diagnostics.Ignore_this_error_message),
+                changes: [createFileTextChanges(sourceFile.fileName, [getIgnoreCommentLocationForLocation(sourceFile, span.start, newLineCharacter)])],
+                groupId,
+            },
+            {
+                description: getLocaleSpecificMessage(Diagnostics.Disable_checking_for_this_file),
+                changes: [createFileTextChanges(sourceFile.fileName, [{
+                    span: {
+                        start: sourceFile.checkJsDirective ? sourceFile.checkJsDirective.pos : 0,
+                        length: sourceFile.checkJsDirective ? sourceFile.checkJsDirective.end - sourceFile.checkJsDirective.pos : 0
+                    },
+                    newText: `// @ts-nocheck${newLineCharacter}`
+                }])],
+                // groupId unnecessary because adding `// @ts-nocheck` even once will ignore every error in the file.
+                groupId: undefined,

```

Comment:
would it be better to split this CodeFix into two, and make sure that a CodeFix is a union type, that either it has fillAllInGroup and returns groupId/groupIds or it does not.

Reply:
I think it would be a pain to have to split up a codefix just to allow multiple (or optional) groupIds. If anything I think we should consolidate them *more*, like `fixClassDoesntImplementInheritedAbstractClass` and `fixClassIncorrectlyImplementsInterface`.